File: typescript/src/tools/tool.ts
export interface Tool {
  readonly name: string;
  readonly description: string;
  execute(args: Record<string, unknown>): Result;
}

export interface ToolExample {
  input: Record<string, unknown>;
  output: string;
  explanation?: string;
}

export interface ExtendedTool extends Tool {
  inputSchema(): Record<string, unknown>;
  examples(): ToolExample[];
}

export class BaseTool implements Tool {
  readonly name: string;
  readonly description: string;

  constructor(name: string, description: string) {
    this.name = name;
    this.description = description;
  }

  execute(_args: Record<string, unknown>): Result {
    throw new Error("Subclass must implement execute");
  }
}

export class Result {
  success: boolean;
  data: unknown = null;
  error: string | null = null;
  timestamp: Date;

  constructor(success: boolean, data?: unknown, error?: string) {
    this.success = success;
    this.data = data ?? null;
    this.error = error ?? null;
    this.timestamp = new Date();
  }

  toJSON(): object {
    return {
      success: this.success,
      data: this.data,
      error: this.error,
      timestamp: this.timestamp.toISOString(),
    };
  }
}

File: python/src/pedro_agentware/tools/tool.py
"""Tool definitions and result types."""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Protocol, runtime_checkable


@runtime_checkable
class Tool(Protocol):
    """Protocol for tool implementations."""

    @property
    def name(self) -> str:
        """Tool name."""
        ...

    @property
    def description(self) -> str:
        """Tool description."""
        ...

    def execute(self, args: dict[str, Any]) -> "Result":
        """Execute the tool with given arguments."""
        ...


@dataclass
class ToolExample:
    """Example usage of a tool."""

    input: dict[str, Any]
    output: str
    explanation: str | None = None


class BaseTool(ABC):
    """Base class for tools."""

    @property
    @abstractmethod
    def name(self) -> str:
        """Tool name."""
        pass

    @property
    @abstractmethod
    def description(self) -> str:
        """Tool description."""
        pass

    def execute(self, args: dict[str, Any]) -> "Result":
        """Execute the tool. Override in subclass for custom logic."""
        raise NotImplementedError("Subclass must implement execute")


@dataclass
class Result:
    """Result of tool execution."""

    success: bool
    data: Any = None
    error: str | None = None
    timestamp: datetime = field(default_factory=datetime.now)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "success": self.success,
            "data": self.data,
            "error": self.error,
            "timestamp": self.timestamp.isoformat(),
        }
